packages/extension-commerce-bm-seo/src/setup-app.tsx

```diff
@@ -0,0 +1,181 @@
+                component,
+                componentProps: props,
+                exact: true
+            }
+        ])
+    }
+
+    /**
+     * TODO: update comment to make it clearer that beforeRouteMatch is also called on the client side
+     * This method is used on the server during the rendering pipeline. It's provided a list of all the routes that your application
+     * is configured with, including those defined in the base application and those added by all the extensions. You can use this
+     * method to modify these routes in any way you want, but you must return an array of routes as a result.
+     */
+    beforeRouteMatch({allRoutes}: BeforeRouteMatchParams): RouteProps[] {
+        const updatedRoutes = [...allRoutes]
+
+        const getComponentName = (displayName: string) => displayName.split('.').pop()
+
+        const findComponentByName = (name: string): React.ComponentType<any> | undefined =>
+            updatedRoutes.find((r) => r.component?.displayName?.includes(name))?.component
+
+        // TODO: do we need to loop through all the possible localized routes?
+        for (const route of updatedRoutes) {
+            if (!route.component.isPlaceholder) continue
+
+            const {displayName} = route.component as any // TODO: fix the type
+            const componentProps = route.componentProps || {}
+            if (!displayName) continue
+
+            const componentName = getComponentName(displayName)
+            if (!componentName) continue
+
+            const actualComponent = findComponentByName(componentName)
+            if (!actualComponent) {
+                throw new Error(`Could not find component with displayName "${componentName}"`)
+            }
+
+            // We need to wrap the component with withPropsWrapper again to ensure that the props
+            // are passed correctly and static react methods are copied over.
+            route.component = withPropsWrapper(actualComponent, componentProps)
+        }
+
+        return updatedRoutes
+    }
+
+    getComponentMap(): ComponentMap {
+        // This extension handles deserialization of routes in beforeRouteMatch instead
+        // of using the component map. This is because this extension relies on components
+        // defined in other extensions which can only be resolved in beforeRouteMatch
+        // where all routes are available.
+
+        // TODO: use resourceTypeToComponentMap to resolve components
+        const {resourceTypeToComponentMap} = this.getConfig() as {
+            resourceTypeToComponentMap: Record<string, string>
+        }
+
+        // Create a map of the component names to a placeholder component for each unique name
+        return Array.from(new Set(Object.values(resourceTypeToComponentMap))).reduce(
+            (acc: ComponentMap, name) => {
+                acc[name] = createPlaceholderComponent(name)
+                return acc
+            },
+            {} as ComponentMap
+        )
```

Comment:
This will create a map where the key's are the component names referenced in `resourceTypeToComponentMap` and the values are a placeholder component that has the displayName set to the component name

E.g., 
```
{
  ProductList: createPlaceholderComponent('ProductList'),
  ProductDetail: createPlaceholderComponent('ProductDetail'),
}
```